File: Data Structures & Algorithms/sliding-window-maximum/submission-0.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        prefix = [0] * n
        suffix = [0] * n

        prefix[0] = nums[0]
        suffix[n - 1] = nums[n - 1]

        for i in range(1, n):
            prefix[i] = nums[i] if i % k == 0 else max(nums[i], prefix[i - 1])
            j = n - i - 1
            suffix[j] = nums[j] if j % k == 0 else max(nums[j], suffix[j + 1])

        result = []
        for i in range(n - k + 1):
            result.append(max(suffix[i], prefix[i + k - 1]))
        
        return result

File: Data Structures & Algorithms/find-pivot-index/submission-0.java
class Solution {
    public int pivotIndex(int[] nums) {
        // 1, 7, 3, 6, 5, 6
        // 1  8 11 18 23 29
        int[] prefix = new int[nums.length + 1];
        for (int i = 1; i < prefix.length; i++) {
            prefix[i] = prefix[i - 1] + nums[i - 1];
        }

        int r = prefix.length - 1;
        for (int i = 1; i < prefix.length; i++) {
            int rightSideSum = prefix[r] - prefix[i];
            int leftSideSum = prefix[i - 1];
            if (leftSideSum == rightSideSum) {
                return i - 1;
            }
        }

        return -1;

    }
}